### PR TITLE
166: Age/Gender Recognition Demo

### LINKED ISSUE #166: Age/Gender recognition demo
We need a technical demo with advanced Python-based pre-/post-processing features to refer to it from the documentation and to address user demands.

### PR DESCRIPTION
**Pull request by dorgun**

### CODE REVIEW

File: samples/age_gender_recognition/README.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
+# People detection, tracking and face blurring (PeopleNet, Nvidia Tracker, OpenCV CUDA)
```
Comment (bwsw) [inline samples/age_gender_recognition/README.md]: Bug!

File: samples/age_gender_recognition/README.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
+# Faces detection, tracking and age-gender recognition (YoloV5face, Nvidia Tracker, Age-Gender model)
+
+A pipeline that uses [Yolov5face](https://github.com/deepcam-cn/yolov5-face) model to detect faces. 
+The model detected faces and 5 landmarks (eyes, nose, mouth). Landmarks are used
```
Comment (abramov-oleg) [inline samples/age_gender_recognition/README.md]:
```suggestion
A pipeline that uses [Yolov5face](https://github.com/deepcam-cn/yolov5-face) model to detect faces and 5 face landmarks (eyes, nose, mouth). Landmarks are used 
```

File: samples/age_gender_recognition/README.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
+# Faces detection, tracking and age-gender recognition (YoloV5face, Nvidia Tracker, Age-Gender model)
+
+A pipeline that uses [Yolov5face](https://github.com/deepcam-cn/yolov5-face) model to detect faces. 
+The model detected faces and 5 landmarks (eyes, nose, mouth). Landmarks are used 
+to calculate the face orientation and preprocessing face image for age/gender model.
```
Comment (abramov-oleg) [inline samples/age_gender_recognition/README.md]:
```suggestion
to calculate the faces orientation and preprocess face images for age/gender model.  
```

File: samples/age_gender_recognition/README.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
+# Faces detection, tracking and age-gender recognition (YoloV5face, Nvidia Tracker, Age-Gender model)
+
+A pipeline that uses [Yolov5face](https://github.com/deepcam-cn/yolov5-face) model to detect faces. 
+The model detected faces and 5 landmarks (eyes, nose, mouth). Landmarks are used 
+to calculate the face orientation and preprocessing face image for age/gender model.  
+Age/gender model estimate age and gender for each face and add this information to
```
Comment (abramov-oleg) [inline samples/age_gender_recognition/README.md]:
```suggestion
Age/gender model estimates age and gender for each face and add this information to 
```

File: samples/age_gender_recognition/README.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
+# Faces detection, tracking and age-gender recognition (YoloV5face, Nvidia Tracker, Age-Gender model)
+
+A pipeline that uses [Yolov5face](https://github.com/deepcam-cn/yolov5-face) model to detect faces. 
+The model detected faces and 5 landmarks (eyes, nose, mouth). Landmarks are used 
+to calculate the face orientation and preprocessing face image for age/gender model.  
+Age/gender model estimate age and gender for each face and add this information to 
+the object metadata. The pipeline uses Nvidia Tracker to track face.
```
Comment (abramov-oleg) [inline samples/age_gender_recognition/README.md]:
```suggestion
the object metadata. The pipeline uses Nvidia Tracker to track faces.
```

File: samples/age_gender_recognition/module.yml [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    width: 1280
+    height: 720
+    padding:
+      # Paddings are kept on the output frame
+      keep: true
+      left: 0
+      right: 0
+      # Padding size is 180px on the top of the frame
+      top: 0
+      bottom: 0
```
Comment (abramov-oleg) [inline samples/age_gender_recognition/module.yml]: No reason to specify 0 paddings.

Same for the other module config.

File: samples/age_gender_recognition/module.yml [hunk 7]
```diff
@@ -0,0 +1,118 @@
+name: ${oc.env:MODULE_NAME, 'traffic-meter'}
```
Comment (abramov-oleg) [inline samples/age_gender_recognition/module.yml]: Wrong module name, same for the other module config file.

File: savant/converter/yolo_v5face.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+                selected_preds[:, :4],
+                selected_preds[:, 4],
+                self.nms_iou_threshold,
+            )
+            selected_nms_prediction = selected_preds[keep == 1]
+            xywh = selected_nms_prediction[:, :4]
+            conf = selected_nms_prediction[:, 4:5]
+            class_num = np.expand_dims(
+                np.array([0] * len(conf), dtype=np.float32), axis=1
+            )
```
Comment (abramov-oleg) [inline savant/converter/yolo_v5face.py]:
```suggestion
            class_num = np.zeros_like(conf, dtype=np.float32)
```

File: samples/age_gender_recognition/age_gender_preproc.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+        cuda_stream: cv2.cuda.Stream,
+    ) -> GPUImage:
+        """Aligned and crop face image from frame by applying affine transformation."""
+
+        crop_size = (FACE_WIDTH, FACE_HEIGHT)
+
+        lanbdmarks = object_meta.get_attr_meta(
+            element_name=MODEL_NAME, attr_name='landmarks'
+        ).value
+        face_landmarks = np.array(lanbdmarks).reshape(-1, 5, 2)
```
Comment (abramov-oleg) [inline samples/age_gender_recognition/age_gender_preproc.py]:
```suggestion
        landmarks = object_meta.get_attr_meta(
            element_name=MODEL_NAME, attr_name='landmarks'
        ).value
        face_landmarks = np.array(landmarks).reshape(-1, 5, 2)
```

File: samples/age_gender_recognition/age_gender_preproc.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+        face_img = cv2.cuda.GpuMat(size=crop_size, type=frame_image.gpu_mat.type(), s=0)
+        src_pts = np.float32(face_landmarks)
```
Comment (abramov-oleg) [inline samples/age_gender_recognition/age_gender_preproc.py]: Variable seems to be redundant.

File: samples/age_gender_recognition/docker-compose.l4t.yml [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+      - "888:888"    # HLS
+      - "8889:8889"  # WebRTC
+    volumes:
+      - zmq_sockets:/tmp/zmq-sockets
+      - ../assets/stub_imgs:/stub_imgs
+    environment:
+      - ZMQ_ENDPOINT=sub+connect:ipc:///tmp/zmq-sockets/output-video.ipc
+      - SOURCE_ID=video
+      - FRAMERATE=25/1
+      - STUB_FILE_LOCATION=/stub_imgs/smpte100_1280x900.jpeg
```
Comment (abramov-oleg) [inline samples/age_gender_recognition/docker-compose.l4t.yml]: This module doesn't have a top padding, so it's better to use 1280x720 stub.

Same for the other docker-compose file

**Review by abramov-oleg — CHANGES_REQUESTED** [review]

File: samples/age_gender_recognition/module.yml (hunk 7 above)
Comment (dorgun) [reply samples/age_gender_recognition/module.yml]: fixed

File: samples/age_gender_recognition/docker-compose.l4t.yml (hunk 11 above)
Comment (dorgun) [reply samples/age_gender_recognition/docker-compose.l4t.yml]: fixed

File: samples/age_gender_recognition/age_gender_preproc.py (hunk 10 above)
Comment (dorgun) [reply samples/age_gender_recognition/age_gender_preproc.py]: fixed

File: samples/age_gender_recognition/module.yml (hunk 6 above)
Comment (dorgun) [reply samples/age_gender_recognition/module.yml]: fixed

File: samples/age_gender_recognition/README.md (hunk 1 above)
Comment (dorgun) [reply samples/age_gender_recognition/README.md]: Fixed

**Review by abramov-oleg — APPROVED** [review]

**Pull request merged**